Paper: Structural Cohesion: Visualization and Heuristics for Fast Computation with NetworkX and matplotlib (Torrents)

jtorrents opened the pull request:
Dear reviewers,

A bit of context on this paper. It is based on a chapter of my thesis (see http://arxiv.org/abs/1503.04476 for the full version). I've submitted the full version to Sociology journals (rejected in one, another version is under revision in another now) and some reviewers suggested cutting or shortening the parts related to discussing algorithms, heuristics and implementation details. So I'm splitting the chapter in two publications, the paper submitted here aims to focus precisely on this. However, I felt that it was worth to discuss at length the structural cohesion model of social cohesion because it's not well known outside Sociology, and I think it might be of interest as it offer an alternative approach to meso level network structure than the more common community analysis. This makes the paper a bit too long.

I've implemented the algorithms and heuristics discussed in the paper on top of NetworkX; they will be part of the upcoming 1.10 release. I'm not sure if it's OK to reference pull requests in the paper. I started adding references (ie pull requests URLs) to footnotes, but I thought that it would be better  referencing the official NetworkX documentation instead, however this will have to wait until the 1.10 release for stable URLs.

Finally, the two figures for this paper (one with 4 subfigures, and the other with 6) should ideally be each a whole page with only one column. Is that possible? I've prepared two pdf for each figure (using the `subfig` latex package) and just added them as a regular figure. Now they are too small as they are squashed to one of the two columns. The pdf for the 3d scatter plots is quite big (20MB), I also added small png versions of each subplot in the paper's directory.

Looking forward to your comments

[Edit by jaberg] http://nipy.bic.berkeley.edu:5000/download/16

mdboom commented:
This was a very interesting paper.  I have only a rudimentary knowledge of graph theory, with some work in graph layout algorithms in particular, but knew nothing of structural cohesion definitions before this paper, so I actually appreciated the level of detail in this paper, but I have some suggestions to improve the ordering and "skimmability" of the paper so that the downsides of such a long paper are minimized.  Unless the editors are concerned about length, I'm inclined to keep the length as-is (I, for one, found it well-written and helpful to someone on the periphery of the subject matter).  I think you are likely to have (at least) two classes of readers, with some overlap: (a) those interested in the graph theory and algorithms as they may apply to their own research, and (b) those interested in the results of the study, which would be interesting to anyone doing scientific collaboration -- probably everyone at SciPy.  I think one can get the latter information out of this paper by skipping the "model", "algorithms" and "heuristics" section and I think some text at the end of the introduction to that effect would get the latter audience to get to that stuff at the end of a long paper without skipping it altogether.

In the section "The structural cohesion model", the last three paragraphs discuss applications and applicability of the models, which are probably the most interesting bits for readers who aren't familiar with these tools but who may want to decide if they are applicable to their own work.  I would move these to the beginning of the section and then add some connective tissue along the lines of "The rest of this section discusses various alternative definitions of structural cohesion models."

I think the section "Structural cohesion in collaboration networks" should be renamed "Case study: Structural cohesion in collaboration networks", to make it clear that you are switching gears from the graph theory to an actual application.

In this section, you clearly outline how the data was obtained (great!).  This data all exists publically as-is, but is the extracted/collated data also available somewhere?

Other than that, my comments are pretty low-level.

- [ ] Make sure all variable names are italicized, particularly the inconsistent italicization of "k"
- [ ] Fix the inconsistent hyphens in terms like "k-connected" and "k-component" etc.
- [ ] Refer to figures as "1(a)" rather than "1 a"
- [ ] I agree with your comment above that the figures need to be full-page.  @katyhuff: Who is a good contact for help with the documentation-production pipeline?

katyhuff commented:
I think you can make the figures at least full-page width using the `:figclass:w` directive attribute. Have you given that a shot @jtorrents ? Alternatively, maybe the `:figclass:p` ? If neither of those works, let me know and we'll see what can be done (anything is possible with enough python?).

jtorrents commented:
@mdboom Thank you for your review and comments. They are really very useful. I'll do all changes that you suggest shortly, I think they will improve the paper and help making it more accessible for SciPy audience. 

Regarding the data and the scripts that do the computations and generate the figures, I'm preparing to publish a git repo with all materials. However, it now also contains other parts of my thesis (which are still work in progress) and will need a bit of clean up before making it publicly available. Maybe the easier approach is to just release the generated networks in a standard format along with this paper. With the networks and the code released with NetworkX 1.10, it is easy to reproduce the analysis in the paper. What do you think?

@katyhuff , thanks! with the directive attribute `:figclass:wp` I got full page figures without two columns. Before I only tried the `p` directive and was getting the figure in a separate page but still with two columns.

I'll push a revised version of the paper shortly.

jtorrents commented:
I've done the suggested changes. I also added footnotes which point to the relevant NetworkX documentation for each concept. They refer to the documentation of the release 1.10, which is imminent. From there it's easy to take a look at the implementation: there is a link to the source code for each function.

I also added a footnote (#10) with links to the networks analysed in the paper in graphml format. However these links point to a currently private github repository, I'm cleaning it up and plan to make it public shortly. However I'm not sure that provide an URL of a github repo is the best way to make the networks available. I also added the graphml files to this PR. Is there a better way to provide these files as a supplementary material of the paper?

katyhuff commented:
This is clearly a very high quality paper, but suffers enormously from too much jargon and detail. While it may have been presented at exactly the right level for the thesis committee of the author (probably experts in the subfield) it is somewhat impenetrable to scientists outside of social network theory. Given the interdisciplinary nature of SciPy, I strongly feel that the authors should make a concerted effort to explain background concepts, neglect unnecessary formal details, and reduce jargon. I spend far too much time searching for definitions of what the author treated as basic concepts (had to google operationalization). In my opinion this paper should be accepted only if:

- the page number is reduced somewhat
- inessential text discussing the "inside baseball" of social network theory leading to the method at hand is reduced significantly
- reliance on domain-specific jargon and unnecessarily complex sentence constructions are severely reduced (a few examples are given below and pointed out inline.)

## Written Quality

The prose is written in english and communicates a narrative. However, the clarity of the narrative is severely hampered by unnecessarily verbose sentence construction and extreme overuse and under-clarification of domain-specific jargon. One of our guidelines to authors and reviewers is that **Due to the interdisciplinary nature of SciPy, highly domain-specific jargon should be avoided or explained where possible.** This paper needs significant edits to meet this requirement.

I think one big fix could be to reduce the introduction to one or two paragraphs (the first one and the last one), follow it up with the essential "Cohesion in Social Networks" section (which actually gives readers outside of your specific subfield a vocabulary for what's going on), and then follow that with the "Structural Cohesion Model". I suggest the author consider *actually just throwing away* the information in the introduction about the structural cohesian model, because it's not understood in the intro and just gets repeated in its own section anyway.

Perhaps the author would benefit by reading this paper aloud to identify and consider simpler constructions, clearer definitions, and more consistent uses of words and phrases like:

- operationalization
- theoretical conceptualization (perhaps "concept" or "theory")
- social cohesion, group cohesion, structural cohesion (are they interchangeable?)
- synthetically communicate (how does one "synthetically communicate?" Certainly you don't mean your plot has a high morpheme-per-word ratio?)
- sociological concept of embeddedness (mentioned once but neither used nor explained)
- collectivity (is this a group? a subgroup? a cohesive group? an uncohesive group?)
- one-mode and two-mode networks (never clarified)
- and many others.

## Technical Content

- The technical content should be scientifically sound.
- Computational content should, likewise, be verified.
- The work should describe the development or use of python software for 
  approaching a problem in a domain within the scope of the conference.

## Novelty and Impact

I am not an expert in the field of social networks, but am sufficiently conversant in graph theory to see that this constitutes a useful domain-specific implementation of a nodal network.

## Verifiability

The author asked, referring to github "Is there a better way to provide these files as a supplementary material of the paper?". The answer is yes. I would recommend using a service like figshare to generate a DOI that persistently points to a specific commit hash or version of your github repository.

## Other Requirements

Good work on the little stuff:
- Code mentioned is cited. 
- Acronyms are avoided where possible and are successfully spelled out upon first mention.

### Length

**The compiled version should be no longer than 8 pages, including figures.** 

At 13 pages, this paper occupies about 160% of the page limit. Notably, the bibliography is substantive and the figures are quite large. However, despite these considerations, I think the paper is generally overdetailed, occasionally redundant, and would be enormously improved by an editorial sweep focused on brevity. I recommend the author reduce the page count by simplifying word choice (why say "theoretical conceptualization" when you mean "theory"?) and removing sidenotes that do not directly contribute to the work at hand.

### Figures

Figures are of high quality and have captions. 

### Code Snippets

- Code snippets should be formatted to fit inside a single column without
  overflow.
- If not in Python, the language in the code snippet should be mentioned.
- Code snippets should follow a common style guide. PEP8 is preferred.

katyhuff commented on papers/jordi_torrents/structural_cohesion.rst:
```diff
@@ -0,0 +1,421 @@
+
+.. class:: abstract
+
+    The structural cohesion model is a powerful sociological conception of cohesion in social groups, but its diffusion in empirical literature has been hampered by operationalization and computational problems. We present useful heuristics for computing structural cohesion that allow a speed-up of one order of magnitude over the algorithms currently available. Both the heuristics and the exact a ...
```
- Is there any toy example of a problem in the world that this implementation helps social scientists to solve? If not, why are you doing it?
- "co-maintenance of Debian packages, co-authorship in Nuclear Theory and High-Energy Theory" is two things. You mean the three following things "co-maintenance of Debian packages, co-authorship in Nuclear Theory, and co-authorship in High-Energy Theory"
- The last sentence could be clearer if split. "We also introduce a novel graphical representation of the structural cohesion analysis to quickly spot differences across networks. It is implemented using matplotlib"

katyhuff commented on papers/jordi_torrents/structural_cohesion.rst:
```diff
@@ -0,0 +1,421 @@
+
+Group cohesion is a central concept that has a long and illustrious history in sociology and organization theory, although its precise characterization has remained elusive. Its use in most sociological research has been ambiguous at best. This is largely because, as [moody2003]_ argued, it is often based on sloppy operationalization grounded mostly in intuition and common sense. Network analysis ...
+
+However, neither the classical approaches nor new developments in community analysis are well-enough suited to address many of the common uses of group cohesion in the sociological and organizational literature, for three key reasons. First, while most of these measures can help us identify cohesive subgroups, they do not provide insight into their robustness, which is a critical element to the t ...
```
"is a critical element to the theoretical conceptualization of cohesion" is a mouthful. Why not aim for clarity:
"is critical to the theory of cohesion"

The great physicist Paul Dirac once said "In science one tries to tell people, in such a way as to be understood by everyone, something that no one ever knew before. But in the case of poetry, it's the exact opposite!"

katyhuff replied:
Also, you have an ambiguous antecedent. What does "its" refer to in this sentence:


> " In most cases, the removal of only a few actors from the subgroups can lead to its fragmentation into smaller disconnected groups"

Is it the group, a particular subgroup?... it can't be "the subgroups" because "the subgroups" plural... so you would need "their fragmentation")

katyhuff commented on papers/jordi_torrents/structural_cohesion.rst:
```diff
@@ -0,0 +1,421 @@
+
+.. [#] See http://networkx.readthedocs.org/en/latest/reference/generated/networkx.algorithms.connectivity.connectivity.node_connectivity.html .
+
+In this paper we extend the structural cohesion model by using the concept of average node connectivity, that is the *average number of actors who need to be removed from the group to disconnect an arbitrary pair of actors in  the group*. We present a set of heuristics to compute structural cohesion based on the fast approximation to compute pairwise node independent paths [white2001]_. We implem ...
```
synthetically communicate? I googled it and this doesn't really make any sense as a construction. It's distracting and makes me wonder what kind of graph has anything to do with linguistic topology (https://en.wikipedia.org/wiki/Synthetic_language). Maybe just say "communicate"?

katyhuff commented on papers/jordi_torrents/structural_cohesion.rst:
```diff
@@ -0,0 +1,421 @@
+
+It offers insight, also, into emerging trust relations among neighborhood residents or the hiring relations among top level US graduate programs [grannis2009]_. In addition to social solidarity and group cohesion, the model can equally fit many relevant theoretical issues, such as conceptualizing structural differences among fields and organizations [white2004]_, operationalizing the structural c ...
+
+The structural cohesion approach to subgroup cohesion ([white2001]_, [moody2003]_) is grounded on two mathematically equivalent definitions of cohesion that are based on commonly used concepts of cohesion in the sociological literature. On the one hand, the ability of a collectivity to hold together independently of the will of any individual. As set out by the formal definition, *a group's struc ...
```
In:

> "Thus Menger's theorem links with an equivalence relation a structural property of graphs ---connectivity based on cut sets--- with how graphs are traversed ---the number of node independent paths among pairs of different nodes."

Are you saying that "how graphs are traversed" = "the number of node-independent paths among pairs of different nodes?

If these are two definitions embedded in one sentence, I recommend simplifying the sentence.

katyhuff commented on papers/jordi_torrents/structural_cohesion.rst:
```diff
@@ -0,0 +1,421 @@
+
+Accuracy can be improved by rebuilding H from the pairwise node connectivity in :math:`G_{candidate}` and following the remaining steps of the heuristics at the cost of slowing down the computation. There is a trade-off between speed and accuracy. After some tests we decided to compute H only once and lean towards the speed pole of the trade-off. Our goal is to have an usable procedure for analyz ...
+
+Our tests reveal that the use of [white2001b]_ approximation does indeed underestimate the order of some :math:`k`-components, particularly in not very sparse networks. One approach to mitigate this problem is to relax the strict cohesion requirement of :math:`H_{candidate}` being a clique. Following the network literature on cliques, we can relax its cohesion requirements in terms of degree, cor ...
```
Consider that "criteria" is technically singular. The plural form is criterion, if we're being formal.

jtorrents commented:
@katyhuff Thanks for you review and sorry for being slow to reply. I started working on the editing/rewriting of the paper following your comments. It's still work in progress but I think I'll be able to meet your requirements. I plan to keep working on this during the weekend. I'll also reply to your comments before the deadline.

Thanks for the pointer to figshare, seems very useful. Instead of generating a DOI that links to a specific commit in my repo, I opted for uploading the network files there and generate an DOI for each. I added to the paper their URL in a footnote in the `Case study` section like this:

• Debian Lenny: http://dx.doi.org/10.6084/m9.figshare.1472938
• Nuclear Theory: http://dx.doi.org/10.6084/m9.figshare.1472940
• High Energy Theory: http://dx.doi.org/10.6084/m9.figshare.1472939 

I prefer this approach so I don't have to clean up the mess in my private github repo to make the files available and can focus on rewriting.

jtorrents commented:
@katyhuff , @jaberg , @mdboom the paper is ready for a decision. I think that the last 4 commits address @katyhuff comments and requirements. Now the paper is 10 pages long, with two full page figures, a small table, 10 footnotes, and references.

As suggested, I removed the redundant parts of the introduction, eliminated the discussion of "inside baseball" of sociology and network theory, removed most of domain-specific jargon and explained the concepts that I thought were necessary. In the case of 1 mode and 2 mode networks, I added a paragraph at the beginning of the `Case Study` section explaining what bipartite and unipartite networks are. In the first version of the paper the synonymous bipartite/2 mode and uniartite/1 mode concepts were used indifferently. In this version I consistently use only the bipartite/unipartite terminology because it's the one used in NetworkX documentation. Finally, I tried to simplify complex sentence constructions and to improve the narrative of the paper. I removed most of the side arguments that related to other work in sociology and network theory which were not essential for the discussion of the structural cohesion model.

katyhuff commented:
@jtorrents , Thanks! I'll chat with @jaberg in Austin about the page limit (which is technically 8 pages). Other than that, I'm satisfied.

jaberg commented:
There are <= 8 pages of text here, so I'd say length is okay. Citations and big clear figures account for 2.5 pages out of the total.

Speaking of figures though, the first figure with the unipartite graph is squeezed vertically so the captions don't show properly, could you adjust the layout before we put the final "ready" tag on this one @jtorrents ?

jtorrents commented:
@jaberg Thanks! I'll update the first figure shortly.

jtorrents commented:
@jaberg I've updated the first figure. The captions show properly now.

jtorrents commented:
@jaberg @katyhuff I will do the conference presentation this afternoon. Can I say that this paper will be in the proceedings? Thanks!

katyhuff commented:
I recognize I've just missed your talk. I hope you just said it has been "submitted," since nothing is officially accepted until it has been merged. However, yes, I do believe we will be merging this very shortly, barring something very unexpected. 

All that said, out of respect for other authors, reviewers, and proceedings co-chairs now and in the future, I do want to urge you to please take our page limit more seriously if you ever submit a scipy paper again. Ignoring those guidelines in this case makes a bad precedent. After all, the page limit (8, including figures) was clear when you submitted a 13 page paper. I am ok making an exception in this case, as it seems to be a quality paper, but would prefer that this not become a trend among authors. (The proceedings are kept within a page limit for many reasons, most important of which is the finite nature of available reviewer-hours.)

I hope your talk went well.

jtorrents commented:
@katyhuff Thanks for the explanation. I would like to apologize for submitting, as a first draft, an unreasonable long and over-detailed paper full of jargon. I really appreciate the time you took to review that first submission, and your detailed comments. That helped me a lot to improve the paper. I'll make sure to adhere strictly to the guidelines in the future. In my field (Sociology) page limits are often soft limits (some top journals don't even have page limits). But I realize that in an interdisciplinary environment such as scipy conference you cannot assume that rules that apply to your discipline will also apply to others. Sorry again for my sloppiness, and thanks for your patience and time.
